samples/apps/copilot-chat-app/SKWebApi/Skills/ChatSkill.cs

```diff
@@ -0,0 +1,354 @@
+        {
+            if (message == null)
+            {
+                continue;
+            }
+
+            var estimatedTokenCount = this.EstimateTokenCount(message.Metadata.Text);
+            if (remainingToken - estimatedTokenCount > 0)
+            {
+                historyText += $"\n{message.Metadata.Text}";
+                remainingToken -= estimatedTokenCount;
+            }
+            else
+            {
+                break;
+            }
+        }
+
+        return $"Chat history:\n{historyText.Trim()}";
+    }
+
+    /// <summary>
+    /// This is the entry point for getting a chat response. It manages the token limit, saves
+    /// messages to memory, and fill in the necessary context variables for completing the
+    /// prompt that will be rendered by the template engine.
+    /// </summary>
+    /// <param name="message"></param>
+    /// <param name="context">Contains the 'tokenLimit' and the 'contextTokenLimit' controlling the length of the prompt.</param>
+    [SKFunction("Get chat response")]
+    [SKFunctionName("Chat")]
+    [SKFunctionInput(Description = "The new message")]
+    [SKFunctionContextParameter(Name = "audience", Description = "The audience the chat bot is interacting with.")]
+    public async Task<SKContext> ChatAsync(string message, SKContext context)
+    {
+        var tokenLimit = SystemPromptDefaults.CompletionTokenLimit;
+        var remainingToken =
+            tokenLimit -
+            SystemPromptDefaults.ResponseTokenLimit -
+            this.EstimateTokenCount(string.Join("\n", new string[]
+                {
+                    SystemPromptDefaults.SystemDescriptionPrompt,
+                    SystemPromptDefaults.SystemResponsePrompt,
+                    SystemPromptDefaults.SystemChatContinuationPrompt
+                })
+            );
+        var contextTokenLimit = remainingToken;
+
+        // Save this new message to memory such that subsequent chat responses can use it
+        try
+        {
+            await this.SaveNewMessageAsync(message, context);
+        }
+        catch (Exception ex) when (!ex.IsCriticalException())
+        {
+            context.Fail($"Unable to save new message: {ex.Message}", ex);
+            return context;
+        }
+
+        // Extract user intent and update remaining token count
+        var userIntent = await this.ExtractUserIntentAsync(context);
+        remainingToken -= this.EstimateTokenCount(userIntent);
+
+        context.Variables.Set("tokenLimit", remainingToken.ToString(new NumberFormatInfo()));
+        context.Variables.Set("contextTokenLimit", contextTokenLimit.ToString(new NumberFormatInfo()));
+        context.Variables.Set("knowledgeCutoff", SystemPromptDefaults.KnowledgeCutoffDate);
+        context.Variables.Set("userIntent", userIntent);
+        context.Variables.Set("audience", context["audience"]);
+
+        var completionFunction = this._kernel.CreateSemanticFunction(
```

Comment:
Same as above, could these be created just once?

Reply:
Let's save this change for fit-and-finish. I'll create a tracking task.